### PR TITLE
[v1.11] IPSec key rotation without agent restart

### PR DESCRIPTION
**Pull request by jibi**
* #19814 -- IPSec key rotation without agent restart (@jibi )
 * #19932 -- ipsec: fix stale keys reclaim logic (@jibi)
 * #20392 -- nodediscovery: make LocalNode return a deep copy of localNode (@jibi)

Moreover, as a dependency of the 2 PRs some additional commits have been backported from https://github.com/cilium/cilium/pull/18887:
  * https://github.com/cilium/cilium/commit/ed934cb958c52515be37998038301229959a0fd2
  * https://github.com/cilium/cilium/commit/9c1cb4856322a80174799dae142ef761e51e2310
  * https://github.com/cilium/cilium/commit/e52fe1d59d1c6f51a76993641ea9412e0f44f749

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 19814 19932 20392 ; do contrib/backporting/set-labels.py $pr done 1.11; done
```
or with
```
$ make add-label BRANCH=v1.11 ISSUES=19814,19932,20392
```

### CODE REVIEW

**Comment by jibi** [comment]
/test-backport-1.11

**Comment by jibi** [comment]
/test-backport-1.11

**Comment by jibi** [comment]
/test-backport-1.11

**Review by julianwiedmann — APPROVED** [review]
ipsec looks good. As expected, code is essentially identical to latest upstream (module the parser diff in `loadIPSecKeys()`).

No surprises in the nodediscovery parts, but hopefully @gandro can have a closer look.

File: pkg/nodediscovery/nodediscovery.go [hunk 1]
```diff
@@ -302,6 +303,16 @@ func (n *NodeDiscovery) UpdateLocalNode() {
...
 }
 
+// LocalNode syncs the localNode object with the information stored in the node
+// package and then returns a copy of the localNode object
+func (n *NodeDiscovery) LocalNode() types.Node {
+	n.localNodeLock.Lock()
+	defer n.localNodeLock.Unlock()
+
+	n.fillLocalNode()
+	return n.localNode
```
Comment (gandro) [inline pkg/nodediscovery/nodediscovery.go]: I didn't notice this in the upstream PR, but don't we want to `DeepCopy()` here? `types.Node` contains a few slices that will be aliased. They are not modified after `StartDiscovery()` has been called in v1.10/v1.11, but they are in v1.12.

**Review by gandro — APPROVED** [review]
Awesome work. Thank you especially for noting merge conflicts in the commits themselves, that was super helpful.

I focused on the nodediscovery part and that looks good (besides one minor thing also present on master). I only did a cursory glance at the IPSec changes.

File: pkg/nodediscovery/nodediscovery.go (hunk 1 above)
Comment (jibi) [reply pkg/nodediscovery/nodediscovery.go]: Makes sense, in that case yes, we should switch to `DeepCopy()` :+1: 

I'll fill the master PR and backport the fix here and in the v1.10 one

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 2]
```diff
@@ -40,6 +40,10 @@ const (
 	IPSecDirOut     IPSecDir = "IPSEC_OUT"
 	IPSecDirBoth    IPSecDir = "IPSEC_BOTH"
 	IPSecDirOutNode IPSecDir = "IPSEC_OUT_NODE"
+
+	// ipSecXfrmMarkSPIShift defines how many bits the SPI is shifted when
+	// encoded in a XfrmMark
+	ipSecXfrmMarkSPIShift = 12
```
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux.go:48]: Commit `ipsec: add ipSecXfrmMark{Set,Get}SPI helpers` is missing a reference to the upstream commit.

**Review by pchaigno — APPROVED** [review]
One minor comment below, can probably be fixed without restarting the end-to-end tests afterward.

**Comment by jibi** [comment]
/test-backport-1.11

**Review by NikAleksandrov — APPROVED** [review]

**Review by qmonnet — APPROVED** [review]

**Pull request merged**